src/aurora.provision-database.ts

```diff
@@ -191,28 +199,47 @@ export async function createUpdate(props: CreateUpdateProps): Promise<awsLambda.
   const m = new Methods();
 
   let clientConfig: ClientConfig;
-  try {
-    console.log('Fetching credentials from Secrets Manager');
-    clientConfig = await m.fetchSecret(managerSecretArn);
-  } catch (err) {
-    return resultFactory({
-      ReasonPrefix: `Secrets issue: ${err}`,
-      Status: CfnStatus.FAILED,
-    });
-  }
-
   let client: Client;
-  try {
-    console.log(`Connecting to database "postgres"`);
-    // While the grants below care which database we are in,
-    // we haven't created the database yet, so we need to connect to postgres.
-    client = await m.connect({ ...clientConfig, database: 'postgres' });
-  } catch (err) {
-    console.log(`connect failed: ${err}`);
-    return resultFactory({
-      ReasonPrefix: `connect failed: ${err}`,
-      Status: CfnStatus.FAILED,
-    });
+
+  let passwordAuthenticationFailedRetries = maxRetries;
+  while (true) {
+    passwordAuthenticationFailedRetries -= 1;
+
+    try {
+      console.log('Fetching credentials from Secrets Manager');
+      clientConfig = await m.fetchSecret(managerSecretArn);
+    } catch (err) {
+      return resultFactory({
+        ReasonPrefix: `Secrets issue: ${err}`,
+        Status: CfnStatus.FAILED,
+      });
+    }
+
+    try {
+      console.log('Connecting to database "postgres"');
+      // While the grants below care which database we are in,
+      // we haven't created the database yet, so we need to connect to postgres.
+      client = await m.connect({ ...clientConfig, database: 'postgres' });
+    } catch (err) {
+      console.log(`connect failed: ${err}`);
+      // If the connect fails because of a 'password authentication failed' error,
+      // We should re-fetch the secret and try connecting again.
+      if (
+        passwordAuthenticationFailedRetries >= 0 &&
+        err instanceof Error &&
+        err.message.includes('password authentication failed')
+      ) {
+        console.log('password authentication failed, refetch password');
+        await wait(retryDelayMs);
+        continue; // Try again
```

Comment:
Not a fan of `continue` and `break` inside `white true`, feels a bit `goto` like 😄. Maybe it would make sense to extract it to a separate function?

Reply:
I agree. I don't love this approach. Can you provide more details around what "extract it to a separate function" would look like?

Reply:
For function sth like this? But this is, I guess, a lot of parameter passing and returning. So in another comment I'll post how it might look without creating a separate function.
```typescript
function sth({ arg1, arg2 }) {
  let passwordAuthenticationFailedRetries = 0;
  let clientConfig;
  let client;

  while (true) {
    try {
      console.log('Fetching credentials from Secrets Manager');
      clientConfig = await m.fetchSecret(managerSecretArn);
    } catch (err) {
      return {
        errResult: resultFactory({
          ReasonPrefix: `Secrets issue: ${err}`,
          Status: CfnStatus.FAILED,
        }),
      };
    }

    try {
      console.log('Connecting to database "postgres"');
      // While the grants below care which database we are in,
      // we haven't created the database yet, so we need to connect to postgres.
      client = await m.connect({ ...clientConfig, database: 'postgres' });

      return { clientConfig, client };
    } catch (err) {
      console.log(`connect failed: ${err}`);
      // If the connect fails because of a 'password authentication failed' error,
      // We should re-fetch the secret and try connecting again.
      if (
        passwordAuthenticationFailedRetries < maxRetries &&
        err instanceof Error &&
        err.message.includes('password authentication failed')
      ) {
        passwordAuthenticationFailedRetries += 1;
        console.log(
          `password authentication failed, refetching password, attempt ${passwordAuthenticationFailedRetries}/${maxRetries} sleeping ${retryDelayMs}`
        );
        await wait(retryDelayMs);
        // rerun
      } else {
        return {
          errResult: resultFactory({
            ReasonPrefix: `connect failed: ${err}`,
            Status: CfnStatus.FAILED,
          }),
        };
      }
    }
  }
}
```

Reply:
Without creating a separate function. Just getting rid of continue and putting break in a place where, I think, is more readable.
```typescript
let passwordAuthenticationFailedRetries = 0;
while (true) {
  try {
    console.log('Fetching credentials from Secrets Manager');
    clientConfig = await m.fetchSecret(managerSecretArn);
  } catch (err) {
    return resultFactory({
      ReasonPrefix: `Secrets issue: ${err}`,
      Status: CfnStatus.FAILED,
    });
  }

  try {
    console.log('Connecting to database "postgres"');
    // While the grants below care which database we are in,
    // we haven't created the database yet, so we need to connect to postgres.
    client = await m.connect({ ...clientConfig, database: 'postgres' });
    break;
  } catch (err) {
    console.log(`connect failed: ${err}`);
    // If the connect fails because of a 'password authentication failed' error,
    // We should re-fetch the secret and try connecting again.
    if (
      passwordAuthenticationFailedRetries < maxRetries &&
      err instanceof Error &&
      err.message.includes('password authentication failed')
    ) {
      passwordAuthenticationFailedRetries += 1;
      console.log(
        `password authentication failed, refetching password, attempt ${passwordAuthenticationFailedRetries}/${maxRetries} sleeping ${retryDelayMs}`
      );
      await wait(retryDelayMs);
      // rerun
    } else {
      return resultFactory({
        ReasonPrefix: `connect failed: ${err}`,
        Status: CfnStatus.FAILED,
      });
    }
  }
}
```

Reply:
Thank you, that's much cleaner!